pkg/gadget-service/instance-manager/file-store/filestore.go

```diff
@@ -0,0 +1,185 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filestore
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/moby/moby/pkg/namesgenerator"
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/protobuf/encoding/protojson"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+	instancemanager "github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/instance-manager"
+)
+
+var GadgetFilePath = "/var/lib/ig"
+
+type FileStore struct {
+	api.GadgetInstanceManagerServer
+	instanceMgr *instancemanager.Manager
+}
+
+func New(mgr *instancemanager.Manager) (instancemanager.Store, error) {
+	fs := &FileStore{
+		instanceMgr: mgr,
+	}
+	err := fs.init()
+	if err != nil {
+		return nil, err
+	}
+	return fs, nil
+}
+
+// init scans GadgetFilePath for existing gadget configurations and runs them.
+func (s *FileStore) init() error {
+	// Load & Apply
+	err := os.MkdirAll(GadgetFilePath, 0o744)
+	if err != nil && errors.Is(err, os.ErrExist) {
+		return fmt.Errorf("creating directory %q: %w", GadgetFilePath, err)
+	}
+
+	return nil
+}
+
+// loadGadgetFile loads a gadget configuration from a file
+func loadGadgetFile(filename string) (*api.InstallGadgetInstanceRequest, error) {
+	// TODO: do we need to sanitize?
+	blob, err := os.ReadFile(filename)
+	if err != nil {
+		return nil, fmt.Errorf("read file %q: %w", filename, err)
+	}
+	gadget := &api.InstallGadgetInstanceRequest{}
+	err = protojson.Unmarshal(blob, gadget)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshal gadget info for file %q: %w", filename, err)
+	}
+	return gadget, nil
+}
+
+// getGadgets returns a list of all installed gadget configurations
+func (s *FileStore) getGadgets() ([]*api.InstallGadgetInstanceRequest, error) {
+	files, err := os.ReadDir(GadgetFilePath)
+	if err != nil {
+		return nil, fmt.Errorf("reading gadgets: %w", err)
+	}
+
+	res := make([]*api.InstallGadgetInstanceRequest, 0)
+	for _, file := range files {
+		if file.IsDir() || !strings.HasSuffix(file.Name(), ".gadget") {
+			continue
+		}
+		gadget, err := loadGadgetFile(filepath.Join(GadgetFilePath, file.Name()))
+		if err != nil {
+			log.Warnf("could not read gadget file: %v", err)
+			continue
+		}
+		res = append(res, gadget)
+	}
+	return res, nil
+}
+
+func (s *FileStore) LoadStoredGadgets() error {
+	gadgets, err := s.getGadgets()
+	if err != nil {
+		return fmt.Errorf("reading existing gadgets: %w", err)
+	}
+
+	for _, gadget := range gadgets {
+		log.Infof("loading gadget instance %q", gadget.GadgetInstance.Id)
+		s.instanceMgr.RunGadget(gadget.GadgetInstance)
+	}
+	return nil
+}
+
+func (s *FileStore) InstallGadgetInstance(ctx context.Context, req *api.InstallGadgetInstanceRequest) (*api.InstallGadgetInstanceResponse, error) {
```

Comment:
I think it's missing a control to avoid multiple gadgets with the same name:


```bash 
$ sudo gadgetctl run --name foo trace_open:latest --detachable
$ sudo gadgetctl list
ID           NAME                                      TAGS                                     GADGET
5eb53da63d99 foo                                                                                trace_open:latest
$ sudo gadgetctl run --name foo trace_open:latest --detachable
$ sudo gadgetctl list
ID           NAME                                      TAGS                                     GADGET
5eb53da63d99 foo                                                                                trace_open:latest
6424b6ae8ca3 foo                                                                                trace_open:latest
```

Reply:
Yeah, I didn't make that unique, yet. The question really is still: who should be in charge here and should the user be able to override the name. Or is name maybe superfluous?